File: client/src/types/events.ts
export type RoleTier = "uninformed" | "informed";

export interface PrimingSource {
  source_type: string;
  name: string;
  tier: "public" | "analytical" | "intelligence";
  lean: "yes" | "no" | "neutral";
  text: string;
}

export interface PrimingBulletin {
  headline: string;
  sources: PrimingSource[];
}

export interface BulletinPayload {
  public: string;
  analytical: string | null;
  intelligence: string | null;
}

export interface RoundStartedEvent {
  round_number: number;
  is_practice_round: boolean;
  round_duration_seconds: number;
  trading_open: boolean;
  current_price: number;
  balance: number;
  yes_held: number;
  no_held: number;
  yes_avg_cost: number | null;
  no_avg_cost: number | null;
  bulletin: BulletinPayload;
  signal_value: "H" | "L" | null;
  signal_theta: number | null;
  round_deadline_unix_ms: number;
}

export interface PriceUpdateEvent {
  current_price: number;
  q_yes: number;
  q_no: number;
  last_trade: {
    participant_id_hashed: string;
    direction: "yes" | "no";
    quantity: number;
    price_before: number;
    price_after: number;
  };
}

export interface MarketOutcomePublicEvent {
  outcome: 0 | 1;
  outcome_label: string;
  true_probability: number;
}

export interface MarketResolvedEvent extends MarketOutcomePublicEvent {
  payout: number;
  final_balance: number;
  pnl: number;
}

export interface RoundEndedEvent {
  round_number: number;
  closing_price: number;
  round_volume: number;
}

export interface ErrorEvent {
  code: string;
  message: string;
}

export interface ParticipantState {
  session_id: number;
  participant_id: string;
  flow_step: string | null;
  phase: string;
  show_tournament_payout_screen: boolean;
  current_market_number: number | null;
  current_round_number: number | null;
  role_tier: RoleTier | null;
  balance: number | null;
  yes_held: number | null;
  no_held: number | null;
  yes_avg_cost: number | null;
  no_avg_cost: number | null;
  current_price: number;
  bulletin: BulletinPayload | null;
  signal_value: "H" | "L" | null;
  signal_theta: number | null;
  is_practice_round: boolean;
  round_duration_seconds: number | null;
  round_deadline_unix_ms: number | null;
  priming: PrimingBulletin | null;
}

export interface ServerToClientEvents {
  state_sync: (payload: ParticipantState) => void;
  market_started: (payload: {
    market_number: number;
    is_practice: boolean;
    stage: number;
    scenario_description: string;
    role_tier: RoleTier;
    endowment_tokens: number;
    starting_balance: number;
    current_price: number;
    max_rounds: number;
    priming: PrimingBulletin | null;
  }) => void;
  round_started: (payload: RoundStartedEvent) => void;
  price_update: (payload: PriceUpdateEvent) => void;
  round_ended: (payload: RoundEndedEvent) => void;
  market_outcome_public: (payload: MarketOutcomePublicEvent) => void;
  market_resolved: (payload: MarketResolvedEvent) => void;
  session_closed: (payload: { session_id: number }) => void;
  error: (payload: ErrorEvent) => void;
}

File: server/events.py
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field


class BulletinPayload(BaseModel):
    public: str
    analytical: str | None = None
    intelligence: str | None = None


class PrimingSourcePayload(BaseModel):
    source_type: str
    name: str
    tier: Literal["public", "analytical", "intelligence"]
    lean: Literal["yes", "no", "neutral"]
    text: str


class PrimingBulletinPayload(BaseModel):
    headline: str
    sources: list[PrimingSourcePayload]


class SessionStartedEvent(BaseModel):
    session_id: int
    session_label: str


class MarketStartedEvent(BaseModel):
    market_number: int
    is_practice: bool = False
    stage: int
    scenario_description: str
    role_tier: Literal["uninformed", "informed"]
    endowment_tokens: float
    starting_balance: float
    current_price: float
    max_rounds: int = 5
    priming: PrimingBulletinPayload | None = None


class RoundStartedEvent(BaseModel):
    round_number: int
    is_practice_round: bool = False
    round_duration_seconds: int = 90
    trading_open: bool = True
    current_price: float
    balance: float
    yes_held: float
    no_held: float
    yes_avg_cost: float | None = None
    no_avg_cost: float | None = None
    bulletin: BulletinPayload
    signal_value: Literal["H", "L"] | None = None
    signal_theta: float | None = None
    round_deadline_unix_ms: int


class LastTradePayload(BaseModel):
    participant_id_hashed: str
    direction: Literal["yes", "no"]
    quantity: int
    price_before: float
    price_after: float


class PriceUpdateEvent(BaseModel):
    current_price: float
    q_yes: float
    q_no: float
    last_trade: LastTradePayload


class RoundEndedEvent(BaseModel):
    round_number: int
    closing_price: float
    round_volume: int


class MarketResolvedEvent(BaseModel):
    outcome: int
    outcome_label: str
    true_probability: float
    payout: float
    final_balance: float
    pnl: float


class MarketOutcomePublicEvent(BaseModel):
    outcome: int
    outcome_label: str
    true_probability: float


class ErrorEvent(BaseModel):
    code: str
    message: str


class TradeRequest(BaseModel):
    side: Literal["buy", "sell"] = "buy"
    direction: Literal["yes", "no"]
    quantity: int = Field(ge=1, le=20)
